File: CipherCore/Sources/CipherCore/Encoding/StegoWire.swift
import Foundation

public enum StegoWire {
    public static let prefix: UInt8 = 0x03
    static let typeMask: UInt8 = 0x0F
    static let deflateFlag: UInt8 = 0x10
    static let paddedFlag: UInt8 = 0x20
    static let knownFlags: UInt8 = typeMask | deflateFlag | paddedFlag

    public static func carriesUnknownFlags(_ payload: Data) -> Bool {
        guard payload.count >= 2, payload.first == prefix else { return false }
        return payload[payload.startIndex + 1] & ~knownFlags != 0
    }

    public static func payloadSize(ciphertext: Int, padded: Bool) -> Int {
        2 + (padded ? Padding.target(4 + ciphertext) : ciphertext)
    }

    public static func fits(ciphertext: Int, padded: Bool) -> Bool {
        payloadSize(ciphertext: ciphertext, padded: padded) <= TextStego.maxPayloadBytes
    }

    public static func frame(_ ciphertext: Data, type: UInt8, deflate: Bool, padded: Bool) -> Data {
        var payload = Data([prefix, (type & typeMask) | (deflate ? deflateFlag : 0) | (padded ? paddedFlag : 0)])
        payload.append(padded ? Padding.frame(ciphertext) : ciphertext)
        return payload
    }

    public static func unframe(_ payload: Data) -> (type: UInt8, deflate: Bool, body: Data)? {
        guard payload.count >= 2, payload.first == prefix else { return nil }
        let flags = payload[payload.startIndex + 1]
        guard flags & ~knownFlags == 0 else { return nil }
        var body = payload.subdata(in: (payload.startIndex + 2) ..< payload.endIndex)
        if flags & paddedFlag != 0 {
            guard let unpadded = Padding.unframe(body) else { return nil }
            body = unpadded
        }
        return (flags & typeMask, flags & deflateFlag != 0, body)
    }
}

File: android/app/src/main/java/com/kryptos/android/core/StegoWire.kt
package com.kryptos.android.core

object StegoWire {
    const val PREFIX = 0x03
    private const val TYPE_MASK = 0x0F
    private const val DEFLATE_FLAG = 0x10
    private const val PADDED_FLAG = 0x20
    private const val KNOWN_FLAGS = TYPE_MASK or DEFLATE_FLAG or PADDED_FLAG

    class Framed(val type: Int, val deflate: Boolean, val body: ByteArray)

    fun carriesUnknownFlags(payload: ByteArray): Boolean {
        if (payload.size < 2 || (payload[0].toInt() and 0xFF) != PREFIX) return false
        return (payload[1].toInt() and 0xFF) and KNOWN_FLAGS.inv() != 0
    }

    fun payloadSize(ciphertext: Int, padded: Boolean): Int =
        2 + (if (padded) Padding.target(4 + ciphertext) else ciphertext)

    fun fits(ciphertext: Int, padded: Boolean): Boolean =
        payloadSize(ciphertext, padded) <= TextStego.MAX_PAYLOAD_BYTES

    fun frame(ciphertext: ByteArray, type: Int, deflate: Boolean, padded: Boolean): ByteArray {
        val body = if (padded) Padding.frame(ciphertext) else ciphertext
        val payload = ByteArray(2 + body.size)
        payload[0] = PREFIX.toByte()
        payload[1] = ((type and TYPE_MASK) or
            (if (deflate) DEFLATE_FLAG else 0) or
            (if (padded) PADDED_FLAG else 0)).toByte()
        body.copyInto(payload, 2)
        return payload
    }

    fun unframe(payload: ByteArray): Framed? {
        if (payload.size < 2 || (payload[0].toInt() and 0xFF) != PREFIX) return null
        val flags = payload[1].toInt() and 0xFF
        if (flags and KNOWN_FLAGS.inv() != 0) return null
        var body = payload.copyOfRange(2, payload.size)
        if (flags and PADDED_FLAG != 0) {
            body = Padding.unframe(body) ?: return null
        }
        return Framed(flags and TYPE_MASK, flags and DEFLATE_FLAG != 0, body)
    }
}
